From a013cb4ea5fd0965bbc7285821d39be72ddefd9a Mon Sep 17 00:00:00 2001
From: Đoàn Trần Công Danh <congdanhqx@gmail.com>
Date: Wed, 3 Sep 2025 10:21:36 +0700
Subject: [PATCH] Qt: always find qt-tools in host environment

When `find_package(Qt6 COMPONENTS <Module>Tools)` is called,
`cmake` will always look into `CMAKE_FIND_ROOT_PATH` first.

Should the `<Module>Tools` in questions happen to be installed in the
build environment (as a dependencies of `<Module>`, `cmake` will pick
that `<Module>Tools` instead of expected host's `<Module>Tools`.

On the other hand, when `cmake` look for Qt components `<Modules>` in
question, it will also look for components `<Modules>Tools` in the
path given by `QT_HOST_PATH` and `QT_HOST_MAKE_CMAKE_DIR` first, then
current `cmake`'s `CMAKE_FIND_ROOT_PATH` later. Thus, we have better
chance to have a runnable tools.

Let's change to look for `<Module>` first, then `<Module>Tools` later,
in order to have better candidate for tools.  The `<Module>Tools`
fallback will be used should `Qt6<Module>` is not available.
---
 docs/CMakeLists.txt               | 1 +
 modules/ECMAddQch.cmake           | 1 +
 modules/ECMFindQmlModule.cmake.in | 1 +
 modules/ECMGenerateQDoc.cmake     | 2 +-
 modules/ECMPoQmTools.cmake        | 1 +
 modules/ECMQueryQt.cmake          | 1 +
 6 files changed, 6 insertions(+), 1 deletion(-)

--- a/docs/CMakeLists.txt
+++ b/docs/CMakeLists.txt
@@ -18,6 +18,7 @@ set_package_properties(
         PURPOSE "Required to build documentation for Extra CMake Modules."
 )
 
+find_package(Qt6 OPTIONAL_COMPONENTS Tools QUIET)
 find_package(Qt6 COMPONENTS ToolsTools)
 set_package_properties(
     Qt6ToolsTools
--- a/modules/ECMAddQch.cmake
+++ b/modules/ECMAddQch.cmake
@@ -458,6 +458,7 @@ function(ecm_add_qch target_name)
             DESCRIPTION "Part of Qt5 tools"
         )
     else()
+        find_package(Qt6 OPTIONAL_COMPONENTS Tools CONFIG QUIET)
         find_package(Qt6 COMPONENTS ToolsTools CONFIG REQUIRED)
         set_package_properties(Qt6ToolsTools PROPERTIES
             TYPE REQUIRED
--- a/modules/ECMFindQmlModule.cmake.in
+++ b/modules/ECMFindQmlModule.cmake.in
@@ -52,6 +52,7 @@ endif()
 # If we haven't checked the version above, use qmlplugindump
 if (NOT CMAKE_CROSSCOMPILING AND NOT MODULE_NOTFOUND AND NOT @GENMODULE@_FOUND AND (NOT DEFINED QT6_IS_SHARED_LIBS_BUILD OR QT6_IS_SHARED_LIBS_BUILD))
     if (QT_MAJOR_VERSION EQUAL "6")
+        find_package(Qt6 OPTIONAL_COMPONENTS Qml QUIET)
         find_package(Qt6 COMPONENTS QmlTools REQUIRED)
         get_target_property(QMLPLUGINDUMP_PROGRAM Qt6::qmlplugindump LOCATION)
     else()
--- a/modules/ECMGenerateQDoc.cmake
+++ b/modules/ECMGenerateQDoc.cmake
@@ -55,7 +55,7 @@ add_custom_target(generate_qch)
 add_custom_target(install_qch_docs)
 
 function(ecm_generate_qdoc target qdocconf_file)
-    find_package(Qt6Tools CONFIG QUIET)
+    find_package(Qt6 OPTIONAL_COMPONENTS Tools CONFIG QUIET)
     find_package(Qt6 OPTIONAL_COMPONENTS ToolsTools CONFIG QUIET)
 
     if (NOT Qt6Tools_FOUND OR NOT Qt6ToolsTools_FOUND)
--- a/modules/ECMPoQmTools.cmake
+++ b/modules/ECMPoQmTools.cmake
@@ -150,6 +150,7 @@ function(ecm_process_po_files_as_qm lang
     if (QT_MAJOR_VERSION EQUAL "5")
         find_package(Qt5LinguistTools CONFIG REQUIRED)
     else()
+        find_package(Qt6 OPTIONAL_COMPONENTS Linguist CONFIG QUIET)
         find_package(Qt6 COMPONENTS LinguistTools CONFIG REQUIRED)
     endif()
 
--- a/modules/ECMQueryQt.cmake
+++ b/modules/ECMQueryQt.cmake
@@ -56,6 +56,7 @@ if (QT_MAJOR_VERSION STREQUAL "5")
     endif()
 elseif(QT_MAJOR_VERSION STREQUAL "6")
     # QUIET to accommodate the TRY option
+    find_package(Qt6 OPTIONAL_COMPONENTS Core QUIET CONFIG)
     find_package(Qt6 COMPONENTS CoreTools QUIET CONFIG)
     set(_exec_name_text "Qt6 qtpaths")
     if (TARGET Qt6::qtpaths)
